Add LoRA linear definition

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #11044

^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

lucylq committed:
Add LoRA linear definition

^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

jackzhxng commented on examples/models/llama/llama_transformer.py:
```diff
@@ -255,7 +256,67 @@ def construct_transformer(model_args: ModelArgs) -> Transformer:
+            and "output_proj" in model_args.target_modules
+            else None
+        )
+        if model_args.attention_type == "static":
```
Nit, should do the wq, wk, wv, wo construction inside the else

JacobSzwejbka commented on examples/models/llama/llama_transformer.py:
```diff
@@ -255,7 +256,67 @@ def construct_transformer(model_args: ModelArgs) -> Transformer:
+            # Static attention constructs ModuleLists for qkvo and
+            # populates them with MHA attention layers; do not pass in
+            # wq, wk, wv, wo.
+            attention = cls(model_args, layer_id, rope)
```
Why are you doing this in llama_transformer at all? You have model args passed to the constructor couldnt you just construct these tensors in the constructor rather then passing them to init?

lucylq replied:
The intention is to make building a transformer composable, e.g., you can pass in any type of attention, with any type of linears.

If construct them inside the transformer constructor, then the transformer has to understand what all the different options are.

JacobSzwejbka replied:
Attention already has to understand what all the different options are. Thats why model args is passed into it no?

JacobSzwejbka replied:
> Attention already has to understand what all the different options are. Thats why model args is passed into it no?

As in its already possible to generate invalid configs where someone sticks model options in and then uses an attention that doesnt support those options.

JacobSzwejbka replied:
This set up is already dependent on model config though. The lora constructors use tons of args from the config

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]

lucylq committed:
Update on "Add LoRA linear definition"


^
Add lora linear definition. Pull out linears from attention, and allow custom linear (eg. lora linear) to be passed in.

If none, construct linear (current behaviour).

Differential Revision: [D73953776](https://our.internmc.facebook.com/intern/diff/D73953776/)

[ghstack-poisoned]
